lib/io/adc.cpp

```diff
@@ -21,26 +20,32 @@ Adc::~Adc()
 
 std::optional<uint16_t> Adc::read()
 {
-  const std::optional<uint16_t> val = resetAndRead4(file_);
-  if(val) 
+  const std::optional<uint16_t> raw_voltage = resetAndRead4(file_);
+  if(raw_voltage) 
   {
       /* TODO: log the value */ 
-      return *val;
+      return *raw_voltage;
   }
   return std::nullopt; 
 }
 
-std::optional<uint16_t> resetAndRead4(int file_descriptor)
+std::optional<uint16_t> resetAndRead4(const int file_descriptor)
 {
-  char buf[4];                                                 // buffer size 4 for fs value
-  lseek(file_descriptor, 0, SEEK_SET);                         // reset file pointer
-  int bytes_read = read(file_descriptor, &buf, sizeof(buf));   // actually consume new data, changes value in buffer
-  if(bytes_read != sizeof(buf)) 
+  char read_buffer[4];                                                               // buffer size 4 for fs value
+  const auto offset = lseek(file_descriptor, 0, SEEK_SET);                           // reset file pointer
+  if(offset != 0)
+  {
+    /*TODO: log that we failed in resetting file_descriptor's file offset*/
+    return std::nullopt; 
+  }
+  const int bytes_read = read(file_descriptor, &read_buffer, sizeof(read_buffer));   // actually consume new data, changes value in buffer
+  if(bytes_read != sizeof(read_buffer)) 
   {
     /*TODO: log that we failed in reading 4 bytes or reached EOF*/
-    return std::nullopt;                                       // returning NULL since we did not get any value
+    return std::nullopt;                                                             // returning NULL since we did not get any value
   }
-  return std::atoi(buf);
+  const int raw_voltage = std::atoi(read_buffer); 
```

Comment:
I think your use of `atoi` is wrong altogether. Surely the `read_buffer` doesn't contain an ASCII representation of the value? 

https://en.cppreference.com/w/cpp/string/byte/atoi

Reply:
Its supposed to be in ASCII since we are reading numbers from a file as characters

Reply:
surely the file is just raw binary, though?

Reply:
That's how the interface for ADC works.

```
[root@beaglebone:/sys/devices/]ocp.3/helper.12# cat AIN0
0
[root@beaglebone:/sys/devices/]ocp.3/helper.12# cat AIN0
427
[root@beaglebone:/sys/devices/]ocp.3/helper.12# cat AIN0
746
[root@beaglebone:/sys/devices/]ocp.3/helper.12# cat AIN0
921
```

https://www.linux.com/training-tutorials/how-get-analog-input-beaglebone-black/

Reply:
hmmm, this seems wrong but i don't know how to disprove it